src/domain/core/src/utils/time_source.rs

```diff
@@ -24,46 +28,132 @@ pub trait SystemTimeSource: Send + Sync {
 /////////////////////////////////////////////////////////////////////////////////////////
 
+type AwaitingCallerId = usize;
+type WokeUpCallersIds = Vec<AwaitingCallerId>;
+
+#[derive(Debug)]
+struct AwaitingCaller {
+    id: AwaitingCallerId,
+    time_to_wake_up: Duration,
+    waker_tx: Option<oneshot::Sender<()>>,
+}
+
+#[derive(Debug, Default)]
+struct SystemTimeSourceStubState {
+    t: Option<DateTime<Utc>>,
+    next_caller_id: AwaitingCallerId,
+    awaiting_callers: Vec<AwaitingCaller>,
+}
+
 #[derive(Debug, Clone)]
 pub struct SystemTimeSourceStub {
-    t: Arc<Mutex<Option<DateTime<Utc>>>>,
+    state: Arc<Mutex<SystemTimeSourceStubState>>,
 }
 
 impl SystemTimeSourceStub {
     pub fn new() -> Self {
         Self {
-            t: Arc::new(Mutex::new(None)),
+            state: Arc::new(Mutex::new(SystemTimeSourceStubState::default())),
         }
     }
 
     pub fn new_set(t: DateTime<Utc>) -> Self {
+        let state = SystemTimeSourceStubState {
+            t: Some(t),
+            next_caller_id: 1,
+            awaiting_callers: vec![],
+        };
+
         Self {
-            t: Arc::new(Mutex::new(Some(t))),
+            state: Arc::new(Mutex::new(state)),
         }
     }
 
     pub fn set(&self, t: DateTime<Utc>) {
-        *self.t.lock().unwrap() = Some(t);
+        let mut state = self.state.lock().unwrap();
+
+        state.t = Some(t);
     }
 
     pub fn unset(&self) {
-        *self.t.lock().unwrap() = None;
+        let mut state = self.state.lock().unwrap();
+
+        state.t = None;
+    }
+
+    pub fn simulate_time_passage(&self, time_quantum: Duration) -> WokeUpCallersIds {
+        let mut state = self.state.lock().unwrap();
+        let mut woke_up_callers_ids = vec![];
+
+        state.awaiting_callers.retain_mut(|awaiting_caller| {
+            awaiting_caller.time_to_wake_up =
+                Duration::zero().max(awaiting_caller.time_to_wake_up - time_quantum);
+
+            let is_time_up = awaiting_caller.time_to_wake_up.is_zero();
+
+            if is_time_up {
+                woke_up_callers_ids.push(awaiting_caller.id);
+
+                let tx = awaiting_caller.waker_tx.take().unwrap();
+
+                tx.send(()).unwrap();
+            }
+
+            !is_time_up
+        });
+
+        if let Some(ref mut t) = state.t {
+            *t += time_quantum;
+        }
+
+        woke_up_callers_ids
     }
 }
 
+#[async_trait::async_trait]
 impl SystemTimeSource for SystemTimeSourceStub {
     fn now(&self) -> DateTime<Utc> {
-        match *self.t.lock().unwrap() {
+        let state = self.state.lock().unwrap();
+
+        match state.t {
             None => Utc::now(),
             Some(ref t) => *t,
         }
     }
+
+    async fn wake_after(&self, duration: Duration) {
+        let (tx, rx) = oneshot::channel();
+
+        {
+            let mut state = self.state.lock().unwrap();
+
+            let new_awaiting_caller = AwaitingCaller {
+                id: state.next_caller_id,
```

Comment:
Why do we need `caller_id`? I don't see it being read anywhere.

Reply:
Ah I see it's used in tests... I wonder if instead of checking the list of caller IDs that were woken up you could `poll()` the actual caller futures and see that only those you expect are ready.

Reply:
If I understand your proposition correctly, it doesn't really work for us, because in `SystemTimeSourceStub::simulate_time_passage()` we mark futures we need as ready.

I mean, we can't check if the futures are ready, because we determine when they are ready

Reply:
Currently your tests do something like this:
```rust
let f = ts.wake_after(10s);
let woken_ids = ts.simulate_time_passage(15s).await;
assert_eq!(woken_ids, [1]);
timeout(f).await.unwrap();
```
What I'm proposing is:
```rust
let f = ts.wake_after(10s);
ts.simulate_time_passage(15s).await;
assert_eq!(f.poll(), Poll::Ready);
```

Now, I'm not 100% sure this will work, but if it does:
- it removes the need for any kind of IDs
- it tells us exactly that this future was completed - you'll be able to remove all the `timeout` logic

Reply:
I see -- a good idea, will check

Reply:
I've experimented with different options, in general, if I allocate more time, it can be done but with ugly workarounds and honestly, not a guaranteed result.

This way we will add a lot of complexity, both in writing/reading for tests (requires `'static` `Future`'s, creating execution contexts for `Future::poll()`, absorbing futures as values in future runners), and in support.

I'd like to suggest keeping the current approach -- agreed that returning IDs is not the most elegant solution, but I think we can afford some liberties for test code

Reply:
Reached out on Slack with some ideas.

For posterity, I managed to express one of the tests like this:
```rust
#[tokio::test]
async fn test_stub_wake_after_with_several_simulate_time_passage_x() {
    let t0 = point_in_time_in_a_parallel_universe();
    let system_time_source = SystemTimeSourceStub::new_set(t0);

    // 1. Start waiting 4 parallel futures:
    //    - a)  60 seconds
    //    - b)  90 seconds
    //    - c) 120 seconds
    //    - d) 120 seconds
    let mut futures = vec![
        Some(system_time_source.wake_after(Duration::seconds(60))),
        Some(system_time_source.wake_after(Duration::seconds(90))),
        Some(system_time_source.wake_after(Duration::seconds(120))),
        Some(system_time_source.wake_after(Duration::seconds(120))),
    ];

    // All futures are not ready
    assert_eq!(
        check_futures_for_completion(&mut futures),
        [false, false, false, false]
    );

    // 2. Simulate +30 seconds:
    //    - a) +00:30/01:00 - waiting
    //    - b) +00:30/01:30 - waiting
    //    - c) +00:30/02:00 - waiting
    //    - d) +00:30/02:00 - waiting
    system_time_source.advance(Duration::seconds(30));
    assert_eq!(
        check_futures_for_completion(&mut futures),
        [false, false, false, false]
    );

    // 3. Simulate +30 seconds:
    //    - a) +01:00/01:00 - done
    //    - b) +01:00/01:30 - waiting
    //    - c) +01:00/02:00 - waiting
    //    - d) +01:00/02:00 - waiting
    system_time_source.advance(Duration::seconds(30));
    assert_eq!(
        check_futures_for_completion(&mut futures),
        [true, false, false, false]
    );

    // 4. Simulate +30 seconds:
    //    - a) +01:30/01:00 - done (before)
    //    - b) +01:30/01:30 - done
    //    - c) +01:30/02:00 - waiting
    //    - d) +01:30/02:00 - waiting
    system_time_source.advance(Duration::seconds(30));
    assert_eq!(
        check_futures_for_completion(&mut futures),
        [true, true, false, false]
    );

    // 4. Simulate +30 seconds:
    //    - a) +01:30/01:00 - done (before)
    //    - b) +01:30/01:30 - done (before)
    //    - c) +01:30/02:00 - done
    //    - d) +01:30/02:00 - done
    system_time_source.advance(Duration::seconds(30));
    assert_eq!(
        check_futures_for_completion(&mut futures),
        [true, true, true, true]
    );
}
```
with this function doing the heavy lifting:
```rust
fn check_futures_for_completion(
    futures: &mut Vec<Option<Pin<Box<dyn Future<Output = ()> + Send + '_>>>>,
) -> Vec<bool> {
    let mut res = Vec::new();
    for i in 0..futures.len() {
        if let Some(f) = &mut futures[i] {
            if f.now_or_never().is_some() {
                res.push(true);
                futures[i] = None;
            } else {
                res.push(false);
            }
        } else {
            res.push(true);
        }
    }
    res
}
```

Reply:
I have my doubts.

Doubt #1: coming back to original example, let''s slightly modify it:

```
let f1 = ts.wake_after(10s); 
let f2 = ts.wake_after(5s);
let woken_ids = ts.simulate_time_passage(15s).await; 
assert_eq!(woken_ids, [2, 1]);
timeout(f1).await.unwrap();
timeout(f2).await.unwrap();
```

The "passed" time jumps too quick, and swallows 2 time slots in the past.
If we allow this as currently, I would expect the awakenings to happen in a certain order that is more or less preserving time relations. Okay, we don't have to restrict the order between 2 futures scheduled after 10s both, but when the time relation is clear, I think we should respect it.

Doubt #2: whether swallowing a timeslot should be allowed at all.
Or, in other words, moving time to a point after an existing scheduled awakening, could be a `panic!`.
If something is planned in 5s, maybe it should be activated exactly at 5s moment only.
Since we are talking about the tests, not the real time, we control the time granularity very well.
Notice the code in flow system that does time rounding and respects minimal scheduler step.
I think it a reasonable delay vs step time is choosen, this could be very accurate and effecient.

Reply:
> use check_futures_for_completion()

thanks for a great tip -- done

Reply:
> If we allow this as currently, I would expect the awakenings to happen in a certain order that is more or less preserving time relations. Okay, we don't have to restrict the order between 2 futures scheduled after 10s both, but when the time relation is clear, I think we should respect it.

this functionality (respect for order) is implemented, but without returning IDs

> Doubt 2: whether swallowing a timeslot should be allowed at all.

I understand the motivation, but in this test case where we don't know exactly who will call `sleep()` and where and when.
Thus, making timestamp shifts exactly to the next time we need to release an awaiting caller will be very problematic.

I'd like to suggest not making this test implementation so demanding